karapace/protobuf/syntax_reader.py

```diff
@@ -0,0 +1,380 @@
+
+        if ch:
+            if self.peek_char() == ch:
+                self.pos += 1
+                return True
+            return False
+        self.skip_whitespace(True)
+        self.expect(self.pos < len(self.data), "unexpected end of file")
+        return self.data[self.pos]
+
+    def push_back(self, ch: str):
+        """ Push back the most recently read character. """
+        if self.data[self.pos - 1] == ch:
+            self.pos -= 1
+
+    def read_string(self) -> str:
+        """ Reads a quoted or unquoted string and returns it. """
+        self.skip_whitespace(True)
+        if self.peek_char() in ['"', "'"]:
+            return self.read_quoted_string()
+        return self.read_word()
+
+    def read_quoted_string(self) -> str:
+        start_quote = self.read_char()
+        if start_quote not in ('"', "'"):
+            raise IllegalStateException(" quote expected")
+
+        result: list = []
+
+        while self.pos < len(self.data):
+            c = self.data[self.pos]
+            self.pos += 1
+            if c == start_quote:
+                if self.peek_char() == '"' or self.peek_char() == "'":
+                    # Adjacent strings are concatenated.  Consume new quote and continue reading.
+                    start_quote = self.read_char()
+                    continue
+                return "".join(result)
+            if c == "\\":
+                self.expect(self.pos < len(self.data), "unexpected end of file")
+                c = self.data[self.pos]
+                self.pos += 1
+                d: str = {
+                    'a': "\u0007",  # Alert.
+                    'b': "\b",  # Backspace.
+                    'f': "\u000c",  # Form feed.
+                    'n': "\n",  # Newline.
+                    'r': "\r",  # Carriage return.
+                    't': "\t",  # Horizontal tab.
+                    'v': "\u000b",  # Vertical tab.
+                }.get(c)
+                if d:
+                    c = d
+                else:
+                    if c in ['x', 'X']:
+                        c = self.read_numeric_escape(16, 2)
+                    elif ord(c) in range(ord('0'), ord('7') + 1):
+                        self.pos -= 1
+                        c = self.read_numeric_escape(8, 3)
+
+            result.append(c)
+            if c == "\n":
+                self.newline()
+
+        self.unexpected("unterminated string")
+
+    def read_numeric_escape(self, radix: int, length: int) -> str:
+        value = -1
+        end_pos = min_of(self.pos + length, len(self.data))
+        while self.pos < end_pos:
+            digit = hex_digit(self.data[self.pos])
+            if digit == -1 or digit >= radix:
+                break
+
+            if value < 0:
+                value = digit
+            else:
+                value = value * radix + digit
+            self.pos += 1
+
+        self.expect(value >= 0, "expected a digit after \\x or \\X")
+        return chr(value)
```

Comment:
This code is part of parsing library. Number of digits is unknown so we cannot use int() call there as soon as we will receive exception.

Reply:
About encoding - I cannot find info about affect of encoding on python call chr( digits ).

Reply:
> Number of digits is unknown so we cannot use int() call there as soon as we will receive exception.

Isn't that what the arg `length' does?

> In both cases it is unicode representation of least 16 bit of int value.

cool, thanks for checking.

Reply:
Length limits number of chars which must be parsed But in case if non digit met then routine return number which found or exception if no number found. So 0x1\t must return 0x1 but int("0x1\t") return exception. 

syntax_reader.py contain code fully ported from square/wire using their parsing method. It is reason why there no regexps. Because they absent in Kotlin code too. I think optimizing and improving of this part  must be done only in case of performance issues and benchmarks.